File: src/server/scripts/cache/cache_maker.py
import json
import math
import requests
from os.path import join, dirname, abspath
from joblib import Parallel, delayed

dir_path = dirname(abspath(__file__))

with open(join(dir_path, 'muns.json')) as munsJson:
    MUNS = json.load(munsJson)

with open(join(dir_path,'ufs.json')) as ufsJson:
    UFS = json.load(ufsJson)

with open(join(dir_path,'biomas.json')) as biomesJson:
    BIOMES = json.load(biomesJson)

BBOX_BRAZIL =  { "bottom": -33.752081, "left": -73.990450, "top": 5.271841, "right": -28.835908 }

MAX_ZOOM_LEVEL_NONE = [0, 1, 2, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]
MAX_ZOOM_LEVEL = [0, 1, 2, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18];

# PORT 5000 - HOMOLOGAÇÃO
# PORT 3000 - PRODUÇÃO
OWS_URL = 'http://127.0.0.1:5000/ows'
LAYERS = ['pasture_quality']
YEARS = [2010, 2018]
# TYPES = ['none', 'cities', 'states','biomes']
TYPES = ['none']

urls = []

def lonToX(lon, zoom):
    n = math.pow(2, zoom)
    x = math.floor( (lon + 180) / 360 * n )
    return (x)

def latToY(lat, zoom):
    lat_rad = lat * math.pi / 180
    n = math.pow(2, zoom)
    y = math.floor((1 - math.log(math.tan(lat_rad) + 1 / math.cos(lat_rad)) / math.pi) / 2 * n)
    return (y)

def tilesInBbox(bbox, zoom):
    tiles = []
    xMin = lonToX(bbox['left'], zoom)
    xMax = lonToX(bbox['right'], zoom)
    yMin = latToY(bbox['top'], zoom)
    yMax = latToY(bbox['bottom'], zoom)

    i = xMin
    while (i <= xMax):
        j = yMin
        while (j <= yMax):
            tiles.append({ "x" : i, "y" : j, "z" : zoom })
            j = j + 1
        i = i + 1

    return (tiles)

def general():
    for layer in LAYERS:
        for year in YEARS:
            for zoom in MAX_ZOOM_LEVEL_NONE:
                tiles = tilesInBbox(BBOX_BRAZIL, zoom)
                for tile in tiles:
                    url = "%s?layers=%s&MSFILTER=year=%s&mode=tile&tile=%s/%s/%s&tilemode=gmap&map.imagetype=png" % (OWS_URL, layer, year, tile['x'], tile['y'], tile['z'])
                    urls.append(url)
def cities():
    for mun in MUNS:
        for layer in LAYERS:
            for year in YEARS:
                for zoom in MAX_ZOOM_LEVEL_NONE:
                    bbox = {"left": mun['left'], "right": mun['right'], "top": mun['top'], "bottom": mun['bottom']}
                    tiles = tilesInBbox(bbox, zoom)
                    for tile in tiles:
                        url = "%s?layers=%s&MSFILTER=year=%s AND cd_geocmu = '%s'&mode=tile&tile=%s/%s/%s&tilemode=gmap&map.imagetype=png" % (OWS_URL, layer, year, mun['cd_geocmu'], tile['x'], tile['y'], tile['z'])
                        urls.append(url)
def states():
    for uf in UFS:
        for layer in LAYERS:
            for year in YEARS:
                for zoom in MAX_ZOOM_LEVEL_NONE:
                    bbox = {"left": uf['left'], "right": uf['right'], "top": uf['top'], "bottom": uf['bottom']}
                    tiles = tilesInBbox(bbox, zoom)
                    for tile in tiles:
                        url = "%s?layers=%s&MSFILTER=year=%s AND uf='%s'&mode=tile&tile=%s/%s/%s&tilemode=gmap&map.imagetype=png" % (OWS_URL, layer, year, uf['uf'], tile['x'], tile['y'], tile['z'])
                        urls.append(url)

def biomes():
    for biome in BIOMES:
        for layer in LAYERS:
            for year in YEARS:
                for zoom in MAX_ZOOM_LEVEL_NONE:
                    bbox = {"left": biome['left'], "right": biome['right'], "top": biome['top'], "bottom": biome['bottom']}
                    tiles = tilesInBbox(bbox, zoom)
                    for tile in tiles:
                        url = "%s?layers=%s&MSFILTER=year=%s AND bioma='%s'&mode=tile&tile=%s/%s/%s&tilemode=gmap&map.imagetype=png" % (OWS_URL, layer, year, biome['bioma'], tile['x'], tile['y'], tile['z'])
                        urls.append(url)

def processRequests(url):
    try:
        #print(str(round((idx / length * 100), 2)) + "% done.")
        requests.get(url)
    except requests.exceptions.RequestException as e:  # This is the correct syntax
        print(e)

def run():
    for type in TYPES:
        if type == 'none':
            general()
        elif type == 'cities':
            cities()
        elif type == 'states':
            states()
        elif type == 'biomes':
            biomes()
    num_cores = 16
    length = len(urls)
    Parallel(n_jobs=num_cores)(delayed(processRequests)(url) for url in urls)

if __name__ == "__main__":

    run()

File: src/server/scripts/cache/cache_maker.js
const bbox   = { bottom : -33.752081, left : -73.990450, top : 5.271841, right : -28.835908 } //Brazil*/
const ufs    = data_ufs.ufs;
const biomas = data_biomas.biomas;
const layers = ["pasture_quality"];
const years  = [2010, 2018];
let urls     = []

const types  = ['none', 'city', 'state', 'biome'];

for (let type of types) {
	if (type === 'none') {
		for (let layername of layers) {
			for (let year in years) {
				for (let zoom in zoomLevels) {
					let tiles = t.tilesInBbox(bbox, zoom)
					tiles.forEach(function (tile) {
						let url = "ogr2ogr -f GPKG bairros_goiania.gpkg PG:\"dbname='covid19' host='172.18.0.4' port='5432' user='tvi' password='tvi123'\" -sql \"SELECT * FROM bairros_municipios WHERE cd_geocmu = '5208707'\"";
						urls.push(url)
					})
				}
			}
		}
	}
	else if (type === 'city') {
		for (let mun of muns) {
			for (let layername of layers) {
				for (let year in years) {
					for (let zoom in zoomLevels) {
						// const _bbox = {left: mun.left, right: mun.right, top: mun.top, bottom: mun.bottom}
						// let tiles = t.tilesInBbox(bbox, zoom)

						tileGenerator(zoom, mun.bbox)
							.on('data', function (tile) {

								var url = ows_url +"/ows"
									+ "?layers=" + layername
									+ "&mode=tile"
									+ "&tilemode=gmap"
									+ "&map.imagetype=png"
									//+ "&map.imagetype=utfgrid"
									+ "&tile=" + [tile[0], tile[1], tile[2]].join('+')

								url += "&MSFILTER=year=" + years[year] + " AND cd_geocmu = '" + mun.cd_geocmu + "'"
								console.log(url)
								urls.push(url)
							});

						// tiles.forEach(function (tile) {
						// 	var url = ows_url +"/ows"
						// 		+ "?layers=" + layername
						// 		+ "&mode=tile"
						// 		+ "&tilemode=gmap"
						// 		+ "&map.imagetype=png"
						// 		//+ "&map.imagetype=utfgrid"
						// 		+ "&tile=" + [tile.x, tile.y, tile.z].join('+')
						//
						// 	url += "&MSFILTER=year=" + years[year] + " AND cd_geocmu = '" + mun.cd_geocmu + "'"
						//
						// 	urls.push(url)
						// })
					}
				}
			}
		}
	}
	else if (type === 'state') {
		for (let uf of ufs) {
			for (let layername of layers) {
				for (let year in years) {
					for (let zoom in zoomLevels) {
						// const _bbox = {left: uf.left, right: uf.right, top: uf.top, bottom: uf.bottom}
						let tiles = t.tilesInBbox(bbox, zoom)
						tiles.forEach(function (tile) {
							let url = ows_url +"/ows"
								+ "?layers=" + layername
								+ "&mode=tile"
								+ "&tilemode=gmap"
								+ "&map.imagetype=png"
								//+ "&map.imagetype=utfgrid"
								+ "&tile=" + [tile.x, tile.y, tile.z].join('+')

							url += "&MSFILTER=year=" + years[year] + " AND uf = '" + uf.uf + "'"

							// console.log(url)
							urls.push(url)
						})
					}
				}
			}
		}
	}
	else if (type === 'bioma') {
		for (let bioma of biomas) {
			for (let layername of layers) {
				for (let year in years) {
					for (let zoom in zoomLevels) {
						// const _bbox = {left: bioma.left, right: bioma.right, top: bioma.top, bottom: bioma.bottom}
						let tiles = t.tilesInBbox(_bbox, zoom)
						tiles.forEach(function (tile) {

							let url = ows_url +"/ows"
								+ "?layers=" + layername
								+ "&mode=tile"
								+ "&tilemode=gmap"
								+ "&map.imagetype=png"
								//+ "&map.imagetype=utfgrid"
								+ "&tile=" + [tile.x, tile.y, tile.z].join('+')

							url += "&MSFILTER=year=" + years[year] + " AND bioma = '" + bioma.bioma + "'"

							// console.log(url)
							urls.push(url)
						})
					}
				}
			}
		}
	}
}

//
// let requests = [];
// const length = urls.length
// urls.forEach(function (url, index) {
// 	requests.push(function (next) {
// 		console.log((index / length * 100).toFixed(2) + "% done." );
// 		request(url, function (error, response, body) {
// 			next()
// 		});
// 	});
// })
//
// async.parallelLimit(requests, multipleRequests)